include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -401,12 +426,160 @@ namespace PETScWrappers
+    const auto ts_resize_setup =
+      [](TS ts, PetscInt it, PetscReal t, Vec x, PetscBool *flg, void *ctx)
+      -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      auto user = static_cast<TimeStepper *>(ctx);
+
+      VectorType xdealii(x);
+      bool       resize = false;
+
+      const int lineno = __LINE__;
+      const int err    = call_and_possibly_capture_ts_exception(
+        user->prepare_for_coarsening_and_refinement,
+        user->pending_exception,
+        []() -> void {},
+        t,
+        it,
+        xdealii,
+        resize);
+      *flg = resize ? PETSC_TRUE : PETSC_FALSE;
+      if (err)
+        return PetscError(
+          PetscObjectComm((PetscObject)ts),
+          lineno + 1,
+          "prepare_for_coarsening_and_refinement",
+          __FILE__,
+          PETSC_ERR_LIB,
+          PETSC_ERROR_INITIAL,
+          "Failure in ts_resize_setup from dealii::PETScWrappers::TimeStepper");
+
+      PetscFunctionReturn(PETSC_SUCCESS);
+    };
+
+    const auto ts_resize_transfer = [](TS       ts,
+                                       PetscInt nv,
+                                       Vec      vin[],
+                                       Vec      vout[],
+                                       void *   ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      auto user = static_cast<TimeStepper *>(ctx);
+
+      std::vector<VectorType> all_in, all_out;
+      for (PetscInt i = 0; i < nv; i++)
+        {
+          all_in.push_back(VectorType(vin[i]));
+        }
+
+      const int lineno = __LINE__;
+      const int err    = call_and_possibly_capture_ts_exception(
+        user->interpolate,
+        user->pending_exception,
+        []() -> void {},
+        all_in,
+        all_out);
+      if (err)
+        return PetscError(
+          PetscObjectComm((PetscObject)ts),
+          lineno + 1,
+          "interpolate",
+          __FILE__,
+          PETSC_ERR_LIB,
+          PETSC_ERROR_INITIAL,
+          "Failure in ts_resize_transfer from dealii::PETScWrappers::TimeStepper");
+
+      if (all_out.size() != all_in.size())
+        SETERRQ(PetscObjectComm((PetscObject)ts),
+                PETSC_ERR_LIB,
+                "Broken all_out");
+
+      for (PetscInt i = 0; i < nv; i++)
+        {
+          vout[i] = all_out[i].petsc_vector();
+          PetscCall(PetscObjectReference((PetscObject)vout[i]));
+        }
+
+      try
+        {
+          user->setup_callbacks();
+        }
+      catch (...)
+        {
+          SETERRQ(PetscObjectComm((PetscObject)ts),
+                  PETSC_ERR_LIB,
+                  "Exception in setup_callbacks");
+        }
+
+      try
+        {
+          user->setup_algebraic_constraints(all_out[0]);
+        }
+      catch (...)
+        {
+          SETERRQ(PetscObjectComm((PetscObject)ts),
+                  PETSC_ERR_LIB,
+                  "Exception in setup_algebraic_constraints");
+        }
```

Comment:
These two functions (`setup_callbacks()` and `setup_algebraic_constraints()`) are under your own control. Can they throw at all? If not, I see no reason why you need to catch exceptions here. (If these functions call something that may throw, the right place to deal with that is *there*, I think.)

Reply:
I use `AssertPETSc` inside those functions that throws if the returned error code from PETSc is not `PETSC_SUCCESS`